File: algorithms/algorithms-cplusplus/leetcode/Question_0295_Find_Median_from_Data_Stream.cc
#include <queue>

using namespace std;

class MedianFinder {
 public:
  MedianFinder() {
  }
    
  void addNum(int num) {
    if (even) {
      large.push(num);
      small.push(large.top());
      large.pop();
    } else {
      small.push(num);
      large.push(small.top());
      small.pop();
    }
    even = !even;
  }
    
  double findMedian() {
    if (even) {
      return (small.top() + large.top()) / 2.0;
    } else {
      return small.top();
    }
  }

 private:
  priority_queue<int, vector<int>, greater<int>> small;
  priority_queue<int> large;
  bool even = true;
};

File: algorithms/algorithms-java/src/main/java/com/lanceolata/coding/algorithms/java/leetcode/Question_0295_Find_Median_from_Data_Stream.java
package com.lanceolata.coding.algorithms.java.leetcode;

import java.util.Collections;
import java.util.PriorityQueue;
import java.util.Queue;

public class Question_0295_Find_Median_from_Data_Stream {
    private Queue<Integer> small;
    private Queue<Integer> large;
    private boolean even;

    public Question_0295_Find_Median_from_Data_Stream() {
        this.small = new PriorityQueue<>(Collections.reverseOrder());
        this.large = new PriorityQueue<>();
        this.even = true;
    }

    public void addNum(int num) {
        if (even) {
            large.offer(num);
            small.offer(large.poll());
        } else {
            small.offer(num);
            large.offer(small.poll());
        }
        even = !even;
    }

    public double findMedian() {
        if (even) {
            return (small.peek() + large.peek()) / 2.0;
        } else {
            return small.peek();
        }
    }
}
